src/LinearAlgebra/SparseMatrix/CSRMatrix.hpp

```diff
@@ -0,0 +1,737 @@
+    }
+
+    /**
+     * @brief Return the value of a particular element.
+     * @param row The row of the element. This is bounds-checked in debug mode,
+     *            but an out-of-bounds access otherwise is UB.
+     * @param col The column of the elements. Bounds-checked always, since it needs to be searched for.
+     * @return If the element exists in the matrix, returns its value. Otherwise return 0.
+     * @note This access incurs a significant search cost, especially if the matrix is unsorted.
+     */
+    template <class ScalarT, typename IdxT, bool KEEP_SORTED>
+    inline ScalarT CSRMatrix<ScalarT, IdxT, KEEP_SORTED>::valueAt(size_t row, IdxT col) const
+    {
+      // Bounds check for debug builds
+      assert(row < numRows());
+
+      // The indices in the values/col indices arrays that this row starts and ends at.
+      // Our search for the element will be between these two spots
+      IdxT row_start_idx = row_indices_[row];
+      IdxT row_end_idx   = row_indices_[row + 1];
+
+      auto row_start = col_indices_.begin() + row_start_idx;
+      auto row_end   = col_indices_.begin() + row_end_idx;
+
+      // The index in the values/col indices arrays we suspect the element may be found at.
+      size_t found_idx;
+
+      if (isSorted())
+      {
+        // Binary search can be used if the rows are sorted
+        found_idx = std::lower_bound(row_start, row_end, col) - col_indices_.begin();
+      }
+      else
+      {
+        // If the rows aren't sorted, then we must do linear search
+        found_idx = std::find(row_start, row_end, col) - col_indices_.begin();
+      }
+
+      // If the col we're trying to find wasn't found (either beyond the end of the row
+      // or because we found a larger column index), the element isn't present in the
+      // matrix, and we return 0.
+      if (found_idx >= row_indices_[row + 1] || col_indices_[found_idx] != col)
+      {
+        return 0;
+      }
+      else
+      {
+        return values_[found_idx];
+      }
+    }
+
+    /**
+     * @brief A helpful builder utility to help build `CSRMatrix` from their components. When built in debug mode,
+     * enables several checks which ensure that you are building the CSR matrix correctly.
+     * @tparam INCLUDE_DIAGONALS Whether or not we should ensure that empty diagonal elements are created, even if
+     *                           not specified. This way if we axpy with a diagonal matrix later (a common operation
+     *                           with Jacobian matrices) we can do it quickly.
+     * @tparam KEEP_SORTED       Whether or not the sorting of rows should be enforced. If false, then elements in a row
+     *                           can be inserted in arbitrary order.
+     * @tparam USE_TEMPLATE      Whether or not a template matrix should be used. A template matrix should ideally be
```

Comment:
```suggestion
     * @param[in] INCLUDE_DIAGONALS - Whether or not we should ensure that empty diagonal elements are created, even if
     *                           not specified. This way if we axpy with a diagonal matrix later (a common operation
     *                           with Jacobian matrices) we can do it quickly.
     * @param[in] KEEP_SORTED  -     Whether or not the sorting of rows should be enforced. If false, then elements in a row
     *                           can be inserted in arbitrary order.
     * @param[in] USE_TEMPLATE -     Whether or not a template matrix should be used. A template matrix should ideally be
```
Also, why are these capital?

Reply:
These are not parameters to a function (`param`), they are template parameters (`tparam`). As template parameters, they are essentially constants, which the [developer guidelines say to use screaming snake case](https://github.com/ORNL/GridKit/blob/develop/CONTRIBUTING.md#constants).